reset-piv/reset-piv.c

```diff
@@ -0,0 +1,306 @@
+	char one[] = "111111";
+	errf_t *ret = ERRF_OK;
+
+	for (size_t i = 0; i < 4; i++) {
+		ret = piv_change_pin(pt, type, zero, one);
+
+		/*
+		 * If for some reason the pin is '000000' we change the pin
+		 * to something else and retry
+		 */
+		if (ret == ERRF_OK) {
+			zero[0] = '1';
+			continue;
+		}
+
+		if (errf_caused_by(ret, "MinRetriesError") ||
+		    errf_caused_by(ret, "LockError")) {
+			errf_free(ret);
+			return (ERRF_OK);
+		}
+
+		if (errf_caused_by(ret, "PermissionError")) {
+			errf_free(ret);
+			continue;
+		}
+
+		/*
+		 * Currently, pivy's piv_change_pin() doesn't indicate
+		 * if the PIN has been blocked. Instead we have to rely
+		 * on an APDU Error as a proxy. However, a PermissionError
+		 * will also have an APDUError in it's error chain, so this
+		 * check must happen _after_ the PermissionError check.
+		 */
+		if (errf_caused_by(ret, "APDUError")) {
+			errf_free(ret);
+			return (ERRF_OK);
+		}
+
+		return (errf("LockError", ret,
+			    "error while attempting to lock PIV"));
+	}
+
+	return (errf("LockError", ret, "failed to lock PIV"));
+}
+
+static errf_t *
+block_piv(struct piv_token *pt)
+{
+	errf_t *ret = ERRF_OK;
+
+	fprintf(stderr, "Blocking PIN...\n");
+	ret = do_block(pt, piv_token_default_auth(pt));
+	if (ret != ERRF_OK)
+		return (ret);
+
+	fprintf(stderr, "Blocking PUK...\n");
+	return (do_block(pt, PIV_PUK));
+}
+
+int
+main(int argc, char * const * argv)
+{
+	errf_t *ret = ERRF_OK;
+	struct piv_token *pt = NULL;
+	char *resp = NULL;
+	char *dbg = getenv("RESET_PIV_DEBUG");
+	int rc;
+
+	alloc_init();
+
+	rc = bunyan_init(getprogname(), &tlog);
+	if (rc != 0) {
+		errx(EXIT_FAILURE, "failed to initialize bunyan logger: %s",
+		    strerror(rc));
+	}
+
+	rc = bunyan_stream_add(tlog, "stderr",
+	    (dbg == NULL) ? BUNYAN_L_INFO : BUNYAN_L_TRACE, bunyan_stream_fd,
+	    (void *)STDERR_FILENO);
+	if (rc != 0) {
+		errx(EXIT_FAILURE, "failed to add stderr stream logger: %s",
+		    strerror(rc));
+	}
+
+	if (dbg != NULL)
+		piv_full_apdu_debug = B_TRUE;
+
+	rc = SCardEstablishContext(SCARD_SCOPE_SYSTEM, NULL, NULL, &ctx);
+	if (rc != 0) {
+		errx(EXIT_FAILURE, "could not initialize libpcsc: %s",
+		    pcsc_stringify_error(rc));
+	}
+
+	ret = get_piv(argv[1], &pt);
+	if (ret != ERRF_OK) {
+		warnfx(ret, "failed to find PIV token to reset");
+		usage();
+	}
+
+	if (!piv_token_is_ykpiv(pt)) {
+		err(EXIT_FAILURE, "only Yubikeys are currently supported; "
+		    "PIV token is not a Yubikey");
+	}
+
+	fprintf(stderr, "Resetting Yubikey %s (%s)\n",
+	    piv_token_shortid(pt), piv_token_rdrname(pt));
+
+	if (ykpiv_token_has_serial(pt))
+		fprintf(stderr, "Serial #%u\n", ykpiv_token_serial(pt));
+
+	fprintf(stderr, "WARNING: this will completely reset the PIV applet "
+	    "on this Yubikey, erasing all keys and certificates!\n");
+
+	do {
+		resp = getpass("Type 'YES' to continue: ");
+	} while (resp == NULL && errno == EINTR);
+
+	if (resp == NULL || strcmp(resp, "YES") != 0)
+		return (0);
```

Comment:
The reset did not happen, so exit non-zero.

Reply:
Done

Reply:
confirmed